File: manager/src/main/java/moe/shizuku/manager/adapter/HomeAdapter.java
package moe.shizuku.manager.adapter;

import android.content.Context;
import android.os.Process;

import java.util.ArrayList;

import moe.shizuku.api.ShizukuService;
import moe.shizuku.manager.ShizukuManagerSettings;
import moe.shizuku.manager.ShizukuManagerSettings.LaunchMethod;
import moe.shizuku.manager.viewholder.ManageAppsViewHolder;
import moe.shizuku.manager.viewholder.ServerStatusViewHolder;
import moe.shizuku.manager.viewholder.StartAdbViewHolder;
import moe.shizuku.manager.viewholder.StartRootViewHolder;
import moe.shizuku.manager.viewmodel.AppsViewModel;
import moe.shizuku.manager.viewmodel.HomeViewModel;
import moe.shizuku.support.recyclerview.IdBasedRecyclerViewAdapter;
import moe.shizuku.support.recyclerview.IndexCreatorPool;

public class HomeAdapter extends IdBasedRecyclerViewAdapter {

    private HomeViewModel mHomeModel;
    private AppsViewModel mAppsModel;

    public HomeAdapter(Context context, HomeViewModel homeModel, AppsViewModel appsModel) {
        super(new ArrayList<>());

        mHomeModel = homeModel;
        mAppsModel = appsModel;

        updateData(context);

        setHasStableIds(true);
    }

    @Override
    public IndexCreatorPool onCreateCreatorPool() {
        return new IndexCreatorPool();
    }

    public void updateData(Context context) {
        final HomeViewModel vm = mHomeModel;
        final boolean v3 = ShizukuService.pingBinder();

        clear();
        addItem(ServerStatusViewHolder.CREATOR, vm.getServiceStatus(), 0);

        addItem(ManageAppsViewHolder.CREATOR,
                (mAppsModel.getGrantedCount().getValue() != null && mAppsModel.getGrantedCount().getValue().data != null)
                        ? mAppsModel.getGrantedCount().getValue().data : 0, 1);

        if (Process.myUid() / 100000 == 0) {
            boolean root = ShizukuManagerSettings.getLastLaunchMode() == LaunchMethod.ROOT;
            boolean rootRestart = vm.getServiceStatus().getUid() == 0;
            if (v3) {
                try {
                    rootRestart |= ShizukuService.getUid() == 0;
                } catch (Throwable ignored) {
                }
            }

            if (root) {
                addItem(StartRootViewHolder.CREATOR, rootRestart, 3);
                addItem(StartAdbViewHolder.CREATOR, null, 2);
            } else {
                addItem(StartAdbViewHolder.CREATOR, null, 2);
                addItem(StartRootViewHolder.CREATOR, rootRestart, 3);
            }
        } else {

        }

        notifyDataSetChanged();
    }
}

File: manager/src/main/java/moe/shizuku/manager/home/HomeAdapter.kt
package moe.shizuku.manager.home

import android.os.Build
import moe.shizuku.manager.management.AppsViewModel
import moe.shizuku.manager.utils.EnvironmentUtils
import moe.shizuku.manager.utils.UserHandleCompat
import rikka.recyclerview.IdBasedRecyclerViewAdapter
import rikka.recyclerview.IndexCreatorPool
import rikka.shizuku.Shizuku

class HomeAdapter(private val homeModel: HomeViewModel, private val appsModel: AppsViewModel) :
    IdBasedRecyclerViewAdapter(ArrayList()) {

    init {
        updateData()
        setHasStableIds(true)
    }

    companion object {

        private const val ID_STATUS = 0L
        private const val ID_APPS = 1L
        private const val ID_TERMINAL = 2L
        private const val ID_START_ROOT = 3L
        private const val ID_START_WADB = 4L
        private const val ID_START_ADB = 5L
        private const val ID_LEARN_MORE = 6L
        private const val ID_ADB_PERMISSION_LIMITED = 7L
    }

    override fun onCreateCreatorPool(): IndexCreatorPool {
        return IndexCreatorPool()
    }

    fun updateData() {
        val status = homeModel.serviceStatus.value?.data ?: return
        val grantedCount = appsModel.grantedCount.value?.data ?: 0
        val adbPermission = status.permission
        val running = status.isRunning
        val isPrimaryUser = UserHandleCompat.myUserId() == 0

        clear()
        addItem(ServerStatusViewHolder.CREATOR, status, ID_STATUS)

        if (adbPermission) {
            addItem(ManageAppsViewHolder.CREATOR, status to grantedCount, ID_APPS)
            addItem(TerminalViewHolder.CREATOR, status, ID_TERMINAL)
        }

        if (running && !adbPermission) {
            addItem(AdbPermissionLimitedViewHolder.CREATOR, status, ID_ADB_PERMISSION_LIMITED)
        }

        if (isPrimaryUser) {
            val root = EnvironmentUtils.isRooted()
            val rootRestart = running && status.uid == 0

            if (root) {
                addItem(StartRootViewHolder.CREATOR, rootRestart, ID_START_ROOT)
            }

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R || EnvironmentUtils.getAdbTcpPort() > 0) {
                addItem(StartWirelessAdbViewHolder.CREATOR, null, ID_START_WADB)
            }

            addItem(StartAdbViewHolder.CREATOR, null, ID_START_ADB)

            if (!root) {
                addItem(StartRootViewHolder.CREATOR, rootRestart, ID_START_ROOT)
            }
        }
        addItem(LearnMoreViewHolder.CREATOR, null, ID_LEARN_MORE)
        notifyDataSetChanged()
    }
}
